### PR TITLE
[proto] Optimized functional pad op for bboxes + tests

### PR DESCRIPTION
**Pull request by vfdev-5**
Pad:
```
[------- pad_bounding_box cpu BoundingBoxFormat.XYXY --------]
            |  pad_bounding_box_old v2  |  pad_bounding_box v2
1 threads: ---------------------------------------------------
      (4,)  |             80            |           10        
6 threads: ---------------------------------------------------
      (4,)  |             84            |           10        

Times are in microseconds (us).

[------- pad_bounding_box cpu BoundingBoxFormat.XYWH --------]
            |  pad_bounding_box_old v2  |  pad_bounding_box v2
1 threads: ---------------------------------------------------
      (4,)  |             50            |           10        
6 threads: ---------------------------------------------------
      (4,)  |             50            |           10        

Times are in microseconds (us).

[------ pad_bounding_box cpu BoundingBoxFormat.CXCYWH -------]
            |  pad_bounding_box_old v2  |  pad_bounding_box v2
1 threads: ---------------------------------------------------
      (4,)  |             50            |           10        
6 threads: ---------------------------------------------------
      (4,)  |             50            |           10        

Times are in microseconds (us).

[------- pad_bounding_box cuda BoundingBoxFormat.XYXY -------]
            |  pad_bounding_box_old v2  |  pad_bounding_box v2
1 threads: ---------------------------------------------------
      (4,)  |            150            |           48        
6 threads: ---------------------------------------------------
      (4,)  |            147            |           48        

Times are in microseconds (us).

[------- pad_bounding_box cuda BoundingBoxFormat.XYWH -------]
            |  pad_bounding_box_old v2  |  pad_bounding_box v2
1 threads: ---------------------------------------------------
      (4,)  |             91            |           47        
6 threads: ---------------------------------------------------
      (4,)  |             91            |           47        

Times are in microseconds (us).

[------ pad_bounding_box cuda BoundingBoxFormat.CXCYWH ------]
            |  pad_bounding_box_old v2  |  pad_bounding_box v2
1 threads: ---------------------------------------------------
      (4,)  |            91.3           |           48        
6 threads: ---------------------------------------------------
      (4,)  |            92.5           |           47        

Times are in microseconds (us).
```

Crop (replaced clone+inplace with a single op):

```
[-------- crop_bounding_box cpu BoundingBoxFormat.XYXY --------]
            |  crop_bounding_box_old v2  |  crop_bounding_box v2
1 threads: -----------------------------------------------------
      (4,)  |             50             |           10         
6 threads: -----------------------------------------------------
      (4,)  |             50             |           10         

Times are in microseconds (us).

[-------- crop_bounding_box cpu BoundingBoxFormat.XYWH --------]
            |  crop_bounding_box_old v2  |  crop_bounding_box v2
1 threads: -----------------------------------------------------
      (4,)  |             97             |           10         
6 threads: -----------------------------------------------------
      (4,)  |            100             |           10         

Times are in microseconds (us).

[------- crop_bounding_box cpu BoundingBoxFormat.CXCYWH -------]
            |  crop_bounding_box_old v2  |  crop_bounding_box v2
1 threads: -----------------------------------------------------
      (4,)  |            100             |           10         
6 threads: -----------------------------------------------------
      (4,)  |            100             |           10         

Times are in microseconds (us).

[------- crop_bounding_box cuda BoundingBoxFormat.XYXY --------]
            |  crop_bounding_box_old v2  |  crop_bounding_box v2
1 threads: -----------------------------------------------------
      (4,)  |            117             |           47         
6 threads: -----------------------------------------------------
      (4,)  |            118             |           48         

Times are in microseconds (us).

[------- crop_bounding_box cuda BoundingBoxFormat.XYWH --------]
            |  crop_bounding_box_old v2  |  crop_bounding_box v2
1 threads: -----------------------------------------------------
      (4,)  |            195             |           47         
6 threads: -----------------------------------------------------
      (4,)  |            196             |           47         

Times are in microseconds (us).

[------ crop_bounding_box cuda BoundingBoxFormat.CXCYWH -------]
            |  crop_bounding_box_old v2  |  crop_bounding_box v2
1 threads: -----------------------------------------------------
      (4,)  |            290             |           48         
6 threads: -----------------------------------------------------
      (4,)  |            292             |           47         

Times are in microseconds (us).

```

cc @datumbox @bjuncek @pmeier

### CODE REVIEW

**Review by datumbox — APPROVED** [review]
LGTM, thanks!

File: torchvision/prototype/transforms/functional/_geometry.py [hunk 1]
```diff
@@ -750,14 +750,12 @@ def pad_bounding_box(
...
     # this works without conversion since padding only affects xy coordinates
-    bounding_box[..., 0] += left
-    bounding_box[..., 1] += top
     if format == features.BoundingBoxFormat.XYXY:
-        bounding_box[..., 2] += left
-        bounding_box[..., 3] += top
+        pad = torch.tensor([left, top, left, top], device=bounding_box.device)
+    else:
+        pad = torch.tensor([left, top, 0, 0], device=bounding_box.device)
+    bounding_box = bounding_box + pad
```
Comment (pmeier) [inline torchvision/prototype/transforms/functional/_geometry.py:758]: Not blocking: instead of repeating the tensor creation, we could only set the values there:

```suggestion
        pad = [left, top, left, top]
    else:
        pad = [left, top, 0, 0]
    bounding_box = bounding_box + torch.tensor(pad, device=bounding_box.device)
```

File: torchvision/prototype/transforms/functional/_geometry.py [hunk 2]
```diff
@@ -750,14 +750,12 @@ def pad_bounding_box(
...
 
-    bounding_box = bounding_box.clone()
-
     # this works without conversion since padding only affects xy coordinates
-    bounding_box[..., 0] += left
-    bounding_box[..., 1] += top
     if format == features.BoundingBoxFormat.XYXY:
-        bounding_box[..., 2] += left
-        bounding_box[..., 3] += top
+        pad = torch.tensor([left, top, left, top], device=bounding_box.device)
```
Comment (pmeier) [inline torchvision/prototype/transforms/functional/_geometry.py:755]: Should we also set the `dtype` here? Otherwise `pad` will always by `int32`. Depending on the dtype of `bounding_box` two issues can happen here:

1. If the bounding box is floating point or int64, the `+` later on will convert it to the correct type, but this is a wasted conversion. Admittedly, this can hardly regarded as perf improvement since we are talking about tiny tensors here.
2. If the bounding box is of a "smaller" integer type, e.g. int16, the `+` later on will upcast it to `int32` and thus violate our rule that the kernel should retain the dtype. This is not caught by the test suite, because we are only testing 
   https://github.com/pytorch/vision/blob/e64784cdea465d833d9d0f66dc73d7abe217933d/test/prototype_common_utils.py#L397

Comment (pmeier) [reply torchvision/prototype/transforms/functional/_geometry.py:755]: The same also applies to `crop_bounding_box` that was changed in #6881 and is also touched in this PR.

File: torchvision/prototype/transforms/functional/_geometry.py [hunk 3]
```diff
@@ -750,14 +750,12 @@ def pad_bounding_box(
 
     left, right, top, bottom = _parse_pad_padding(padding)
 
-    bounding_box = bounding_box.clone()
-
     # this works without conversion since padding only affects xy coordinates
```
Comment (pmeier) [inline torchvision/prototype/transforms/functional/_geometry.py:753]: This should be moved inside the `else` branch, right? And should probably also be added to `crop_bounding_box` below.

File: torchvision/prototype/transforms/functional/_geometry.py (hunk 2 above)
Comment (vfdev-5) [reply torchvision/prototype/transforms/functional/_geometry.py:755]: Yes, I was thinking also about dtype. In general, yes, we need to convert pad to bounding_box dtype here, in practice my thought was that bboxes are floats so padding will be converted to float. I think bboxes as int16 are rare...
OK, i'll update the PR.

File: torchvision/prototype/transforms/functional/_geometry.py (hunk 3 above)
Comment (vfdev-5) [reply torchvision/prototype/transforms/functional/_geometry.py:753]: I'll remove this comment to avoid confusion

Comment (pmeier) [reply torchvision/prototype/transforms/functional/_geometry.py:753]: Also fine by me.

File: torchvision/prototype/transforms/functional/_geometry.py [hunk 4]
```diff
@@ -768,14 +768,11 @@ def pad_bounding_box(
...
-    # this works without conversion since padding only affects xy coordinates
-    bounding_box[..., 0] += left
-    bounding_box[..., 1] += top
     if format == features.BoundingBoxFormat.XYXY:
-        bounding_box[..., 2] += left
-        bounding_box[..., 3] += top
+        pad = [left, top, left, top]
+    else:
+        pad = [left, top, 0, 0]
+    bounding_box = bounding_box + torch.tensor(pad, dtype=bounding_box.dtype, device=bounding_box.device)
```
Comment (lezcano) [inline torchvision/prototype/transforms/functional/_geometry.py:775]: Why is this (and the substraction below) out-of-place rather than in-place?

Comment (vfdev-5) [reply torchvision/prototype/transforms/functional/_geometry.py:775]: We should avoid modifying input inplace. Previously, we were cloning and doing inplace but it is slower. See https://github.com/pytorch/vision/pull/6881#discussion_r1012228887

Comment (datumbox) [reply torchvision/prototype/transforms/functional/_geometry.py:775]: If the intention is to not modify in place the data, it's faster to do out of place than clone and do inplace.

Comment (vfdev-5) [reply torchvision/prototype/transforms/functional/_geometry.py:775]: @datumbox I was quicker )))

Comment (lezcano) [reply torchvision/prototype/transforms/functional/_geometry.py:775]: fair play :D

Comment (lezcano) [reply torchvision/prototype/transforms/functional/_geometry.py:775]: Also, yes, out of place ops should always be faster than clone + in-place, as in the latter ones you read the data twice.
